### PR TITLE
[Feature][AES-684] Expose named exports at the top level

### PR DESCRIPTION
**Pull request by beau-charman**
This PR aims to address https://aesoponline.atlassian.net/browse/AES-684 by exporting named defaults so that they are exposed at the top level.

```
import { Button, useOnScreen, constants } from 'aesop-gel';
```

Blocking: https://github.com/aesop/aesop-web-ui/pull/3539

Might introduce https://www.npmjs.com/package/@rollup/plugin-multi-entry

**EDIT:** 
Importing the following in local aesop-web-ui
Screen Shot 2021-01-14 at 12 55 01 pm

yields the following 
Screen Shot 2021-01-14 at 12 57 21 pm

### CODE REVIEW

**Comment by beau-charman** [comment]
As the PR stands, all named exports are exposed at the top level. 

This will help in making more declarative imports and reduced continually accessing object methods, which can be unnecessary and slightly taxing, for example replacing something like: 

```
import gel from 'aesop-gel';
... 
const { Podium } = gel.components;
const { useOnScreen } = gel.customHooks;
const { constants } = gel;
```

with: 

```
import { Button, useOnScreen, constants } from 'aesop-gel';
```

However, I was wondering if anyone had thoughts around enforcing a styling of importing to maintain name spacing and clarity of where named exports are coming from. For example: 

```
import { Button, customHooks, constants } from 'aesop-gel';
...
const { useOnScreen } = customHooks;
...
constants.BREAKPOINT
```

and so on.

**Comment by MkMan** [comment]
> This PR aims to address https://aesoponline.atlassian.net/browse/AES-684 by exporting named defaults so that they are exposed at the top level.
> 
> ```
> import { Button, useOnScreen, constants } from 'aesop-gel';
> ```

I love this, it matches what I have seen done in custom library guides ([guide1](https://blog.harveydelaney.com/creating-your-own-react-component-library/#using-the-component-library), [guide2](https://medium.com/@mehrahinam/build-a-private-react-component-library-cra-rollup-material-ui-github-package-registry-1e14da93e790), [guide3](https://medium.com/grandata-engineering/how-i-set-up-a-react-component-library-with-rollup-be6ccb700333), [guide4](https://www.grzegorowski.com/react-typescript-library-rollup-jest-adding-react)).

How does Rollup react* to this? When I tried both named and default exports it ~hurled insults at me~ showed warnings

* Pun intended 😛

**Comment by beau-charman** [comment]
> > This PR aims to address https://aesoponline.atlassian.net/browse/AES-684 by exporting named defaults so that they are exposed at the top level.
> > ```
> > import { Button, useOnScreen, constants } from 'aesop-gel';
> > ```
> 
> I love this, it matches what I have seen done in custom library guides ([guide1](https://blog.harveydelaney.com/creating-your-own-react-component-library/#using-the-component-library), [guide2](https://medium.com/@mehrahinam/build-a-private-react-component-library-cra-rollup-material-ui-github-package-registry-1e14da93e790), [guide3](https://medium.com/grandata-engineering/how-i-set-up-a-react-component-library-with-rollup-be6ccb700333), [guide4](https://www.grzegorowski.com/react-typescript-library-rollup-jest-adding-react)).
> 
> How does Rollup **react** to this? When I tried both named and default exports it ~hurled insults at me~ showed warnings

Could you provide a screengrab or example of this? I don't wish to introduce and issues :(

**Comment by MkMan** [comment]
I wonder if this also has the potential of reducing bundle size in the consuming app as it will only import the components it wants and not all of Gel. @maxlew and I were talking about this before.

**Comment by MkMan** [comment]
> Could you provide a screengrab or example of this? I don't wish to introduce and issues :(

image

**Comment by maxlew** [comment]
> I wonder if this also has the potential of reducing bundle size in the consuming app as it will only import the components it wants and not all of Gel. @maxlew and I were talking about this before.

Yeah I think it'd allow all the tree shaking to work properly as we're not importing the entire library with the `import gel from 'aesop-gel';` line anymore.

**Comment by MkMan** [comment]
Personally, I think it's definitely worthwhile switching to named exports from default exports even if it's a breaking change. We'll just need to coordinate it with Web UI. After that we can stop extracting the css to separate file (in Gel) and each component can just be consumed on its own in **1 line**.

**Comment by beau-charman** [comment]
https://www.npmjs.com/package/@rollup/plugin-multi-entry

**Comment by beau-charman** [comment]
@maxlew , @MkMan , thank you for your feedback. 

I have some more changes that force named exports for commonJS (which seems to fix the `export {...} from '...';` issue as well :))

File: src/components/index.js [hunk 1]
```diff
@@ -1,192 +1,61 @@
...
-  SubNav,
-  Textarea,
-  TextInput,
-  TextOverFullWidthAsset,
-  Transition,
-  TwoColumnLayout,
-  TwoColumnList,
-  Video,
-};
+export { Accordion } from './Accordion';
```
Comment (MkMan) [inline src/components/index.js:1]: 😍

**Review by MkMan — APPROVED** [review]

**Review by maxlew — APPROVED** [review]
Looks good, I didn't realize the export from syntax worked, makes all this files very clean.

File: src/components/index.js (hunk 1 above)
Comment (kevin-ho87) [reply src/components/index.js:1]: I never knew this 'export from' existed

File: src/compositions/index.js [hunk 2]
```diff
@@ -1,11 +1,2 @@
...
-
-export { Footer };
-export { NewsletterSignUp };
-
-/* eslint-disable-next-line import/no-default-export */
-export default {
-  Footer,
-  NewsletterSignUp,
-};
+export { Footer } from './Footer';
```
Comment (zack76) [inline src/compositions/index.js:1]: I thought **Composition** is only used for displayed in storybook. These two won't be actually used anywhere

Comment (beau-charman) [reply src/compositions/index.js:1]: nah, compositions are a collection of shared components. anything just for storybook should just be composed in `.stories.js` files.

Comment (beau-charman) [reply src/compositions/index.js:1]: I can remove the compositions from the main index import / export (my other two compositions are not ready to be exposed yet anyway)

**Comment by MkMan** [comment]
Should the commit where the exports change include **BREAKING CHANGE** in the message to make semantic-release bump the major version upon merging to `master`?

**Comment by beau-charman** [comment]
> Should the commit where the exports change include **BREAKING CHANGE** in the message to make semantic-release bump the major version upon merging to `master`?

Is that a question or a suggestion? 😅  Can I ad that commit message to my final commit or should I redo this commit message https://github.com/aesop/aesop-gel/pull/293/commits/536fb4b07e80641dfd752912eded674ac38b94bb?

**Comment by andrwchang** [comment]
brilliant work here well done @beau-charman

**Pull request merged**